rustls/src/client/tls13.rs

```diff
@@ -580,23 +581,30 @@ impl State<ClientConnectionData> for ExpectCertificateRequest {
             .iter()
             .map(|p| p.0.as_slice())
             .collect::<Vec<&[u8]>>();
-        let maybe_certkey = self
+
+        let client_auth = self
             .config
             .client_auth_cert_resolver
-            .resolve(&canames, &compat_sigschemes);
-
-        let mut client_auth = ClientAuthDetails::new();
-        if let Some(certkey) = maybe_certkey {
-            debug!("Attempting client auth");
-            let maybe_signer = certkey
-                .key
-                .choose_scheme(&compat_sigschemes);
-            client_auth.certkey = Some(certkey);
-            client_auth.signer = maybe_signer;
-            client_auth.auth_context = Some(certreq.context.0.clone());
-        } else {
-            debug!("Client auth requested but no cert selected");
-        }
+            .resolve(&canames, &compat_sigschemes)
+            .and_then(|certkey| {
+                certkey
+                    .key
+                    .choose_scheme(&compat_sigschemes)
+                    .map(|signer| {
+                        debug!("Attempting client auth");
+                        ClientAuthDetails::Verify {
+                            certkey,
+                            signer,
+                            auth_context_tls13: Some(certreq.context.0.clone()),
+                        }
+                    })
+            })
+            .unwrap_or_else(|| {
+                debug!("Client auth requested but no cert selected");
+                ClientAuthDetails::Empty {
+                    auth_context_tls13: Some(certreq.context.0.clone()),
+                }
+            });
```

Comment:
So this could likely be shared with the 1.2 side if it's moved into a constructor?